lib/app/modules/smart/smart_discovery_sequence.rb

```diff
@@ -10,71 +10,119 @@ class SMARTDiscoverySequence < SequenceBase
         * [The OAuth 2.0 Authorization Framework](https://tools.ietf.org/html/rfc6749)
         * [OpenID Connect Core](https://openid.net/specs/openid-connect-core-1_0.html)
-
-       )
-
-      test 'Retrieve Authorization from Well Known endpoint' do
+      )
+
+      REQUIRED_WELL_KNOWN_FIELDS = [
+        'authorization_endpoint',
+        'token_endpoint',
+        'capabilities'
+      ].freeze
+
+      RECOMMENDED_WELL_KNOWN_FIELDS = [
+        'scopes_supported',
+        'response_types_supported',
+        'management_endpoint',
+        'introspection_endpoint',
+        'revocation_endpoint'
+      ].freeze
+
+      test 'Retrieve Configuration from well-known endpoint' do
         metadata do
           id '01'
           link 'http://www.hl7.org/fhir/smart-app-launch/conformance/#using-well-known'
           desc %(
-            The authorization endpoints accepted by a FHIR resource server can be exposed as a Well-Known Uniform Resource Identifier
+            The authorization endpoints accepted by a FHIR resource server can
+            be exposed as a Well-Known Uniform Resource Identifier
           )
         end
 
-        oauth_configuration_url = @instance.url.chomp('/') + '/.well-known/smart-configuration'
-        oauth_configuration_response = LoggedRestClient.get(oauth_configuration_url)
-        assert_response_ok(oauth_configuration_response)
-        oauth_configuration_response_body = JSON.parse(oauth_configuration_response.body)
-
-        assert !oauth_configuration_response_body.nil?, 'No authorization response body available'
-        @well_known_authorize_url = oauth_configuration_response_body['authorization_endpoint']
-        @well_known_token_url = oauth_configuration_response_body['token_endpoint']
-        capabilities = oauth_configuration_response_body['capabilities']
-        register_url = oauth_configuration_response_body['registration_endpoint']
+        well_known_configuration_url = @instance.url.chomp('/') + '/.well-known/smart-configuration'
+        well_known_configuration_response = LoggedRestClient.get(well_known_configuration_url)
+        assert_response_ok(well_known_configuration_response)
+        assert_response_content_type(well_known_configuration_response, 'application/json')
+        assert_valid_json(well_known_configuration_response.body)
+
+        @well_known_configuration = JSON.parse(well_known_configuration_response.body)
+        @well_known_authorize_url = @well_known_configuration['authorization_endpoint']
+        @well_known_token_url = @well_known_configuration['token_endpoint']
+        @instance.update(
+          oauth_authorize_endpoint: @well_known_authorize_url,
+          oauth_token_endpoint: @well_known_token_url,
+          oauth_register_endpoint: @well_known_configuration['registration_endpoint']
+        )
+
+        assert @well_known_configuration.present?, 'No .well-known/smart-configuration body'
```

Comment:
It had to do with guaranteeing that the instance variables were set and @instance was updated based on the contents of the well-known endpoint. Consider the following workflow:

- No body is received, so the assertion fails and the oauth urls aren't set
- Later, the urls are set by the conformance sequence
- This test is rerun, no body is received, the assertion fails, and the sequence continiues with the urls received from the conformance sequence
- The tests continue as though these urls were received from the well-known endpoint, leading to erroneous passes

I think I reorganized some code since I encountered this problem, and this code as is may not work as intended, so I will look into this some more.